pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := copyRoutes(old, new); err != nil {
+		return err
+	}
+	if err := createNewRules(
+		linux_defaults.RulePriorityEgress,
+		old,
+		linux_defaults.RulePriorityEgressv2,
+		new,
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress, old); err != nil {
+		return err
+	}
+	if err := deleteOldRoutes(old); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyRoutes(from, to int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: from,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return err
+	}
+
+	// In ENI mode, we only expect two rules:
+	//  1) Link scoped route with a gateway IP
+	//  2) Default route via gateway IP
+	// We must upsert the link scoped route first which will make the kernel
+	// aware of the gateway IP. After that, then we will upsert the "default
+	// via gatewayIP" route. If we didn't do this, then the kernel would
+	// complain with "Error: Nexthop has invalid gateway." with an errno of
+	// ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := netlink.RouteReplace(&route); err != nil {
+				return fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			break
+		}
+	}
+
+	for _, r := range routes {
+		route := r
+		route.Table = to
+		if err := netlink.RouteReplace(&route); err != nil {
+			return fmt.Errorf("unable to replace route under table ID: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func createNewRules(fromPrio, oldTable, toPrio, newTable int) error {
+	rules, err := findRulesByPriority(fromPrio)
+	if err != nil {
+		return err
+	}
+
+	candidates := make([]netlink.Rule, 0, len(rules))
+	for _, r := range rules {
+		if r.Table == oldTable {
+			candidates = append(candidates, r)
+		}
+	}
+
+	for _, r := range candidates {
+		rule := r
+		rule.Priority = toPrio
+		rule.Table = newTable
+		if err := netlink.RuleAdd(&rule); err != nil {
+			return fmt.Errorf("unable to add new rule: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func deleteOldRules(prio, table int) error {
+	rules, err := findRulesByPriority(prio)
+	if err != nil {
+		return err
+	}
+
+	candidates := make([]netlink.Rule, 0, len(rules))
+	for _, r := range rules {
+		if r.Table == table {
+			candidates = append(candidates, r)
+		}
+	}
+
+	for _, r := range candidates {
+		if err := netlink.RuleDel(&r); err != nil {
+			return fmt.Errorf("unable to delete old rules: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func findRulesByPriority(prio int) ([]netlink.Rule, error) {
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
```

Comment:
I would retrieve the list of rules once in the beginning and then use that list instead of doing more netlink calls. Each of them can fail and you have to handle the error.